### PR TITLE
docstore: how-to, part 2: actions

### PR DESCRIPTION
**Pull request by jba**

### CODE REVIEW

File: internal/website/content/howto/docstore/actions.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
+---
+title: "Actions"
```
Comment (zombiezen) [inline internal/website/content/howto/docstore/actions.md]: Optional: Existing how-to guide titles tend to be a verb phrase like "Publish Messages to a Topic". Consider changing the title to "Reading and Writing Documents" or something similar.

File: internal/website/content/howto/docstore/actions.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+In Docstore, you use actions to read, modify and write documents. You can
+execute a single action, or run multiple actions together in an _action list_.
+
+<!--more-->
+
+## Representing Documents
+
+We'll use a collection with documents represented by this Go struct:
+```
```
Comment (zombiezen) [inline internal/website/content/howto/docstore/actions.md]: nit: blank line before code block and use ```` ```go ```` to highlight with Go syntax. (Here and elsewhere.)

````suggestion

```go
````

File: internal/website/content/howto/docstore/actions.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+We recommend using structs for documents because they impose some structure on
+your data, but Docstore also accepts `map[string]interface{}` values. See
+[the `docstore` package
+documentation](https://godoc.org/gocloud.dev/internal/docstore#hdr-Documents)
+for more information.
+
+The `DocstoreRevision` field holds information about the latest revision of the
+document. We discuss it [below]({{< ref "#docrev" >}}).
+<!-- TODO(jba): make sure the above link works -->
```
Comment (zombiezen) [inline internal/website/content/howto/docstore/actions.md]: It should. (Not sure what your dev workflow is, but you can check the site by running `hugo serve -D` from the `internal/website` directory.)

File: internal/website/content/howto/docstore/actions.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+## Actions and Action Lists
+
+Once you have [opened a collection]({{< ref "open-collection.md" >}}), you
+can call action methods on it. We will use `coll` as the variable holding the collection.
+
+Docstore supports six actions on documents:
+
+- Get retrieves a document.
```
Comment (zombiezen) [inline internal/website/content/howto/docstore/actions.md]: nit: Try using code spans to denote the term (here and below).

```suggestion
- `Get` retrieves a document.
```

File: internal/website/content/howto/docstore/actions.md (hunk 1 above)
Comment (jba) [reply internal/website/content/howto/docstore/actions.md]: Done.

File: internal/website/content/howto/docstore/actions.md (hunk 2 above)
Comment (jba) [reply internal/website/content/howto/docstore/actions.md]: Done.

File: internal/website/content/howto/docstore/actions.md (hunk 4 above)
Comment (jba) [reply internal/website/content/howto/docstore/actions.md]: Done.

File: internal/docstore/example_test.go [hunk 5]
```diff
@@ -278,3 +267,59 @@ func ExampleQuery_Update() {
...
+	pat := &Player{Name: "Pat", Score: 7}
+	if err := coll.Create(ctx, pat); err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println(pat) // memdocstore revisions are deterministic, so we can check the output.
+
+	// Double a player's score. We cannot use Update to multiply, so we use optimistic
+	// locking instead.
+
+	// We may have to retry a few times; put a time limit on that.
```
Comment (shantuo) [inline internal/docstore/example_test.go]: Is this needed even for the memdocstore?

File: internal/website/content/howto/docstore/actions.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+If you have more than one action to perform, it is better to use an action
+list. Drivers can optimize action lists by using bulk RPCs, running the actions
+concurrently, or employing a provider's special features to improve efficiency
+and reduce cost. Here we create several documents using an action list.
+
+{{< goexample "gocloud.dev/internal/docstore.ExampleCollection_Actions_bulkWrite" >}}
+
+`ActionList` has a fluent API, so you can build and execute a sequence of
+actions in one line of code. Here we `Put` a document and immediately `Get` its new
```
Comment (shantuo) [inline internal/website/content/howto/docstore/actions.md]: Should we mention this example may not get the new value depending on whether the service supports it?

File: internal/docstore/example_test.go (hunk 5 above)
Comment (jba) [reply internal/docstore/example_test.go]: Yes, if some other goroutine in the same program is doing the same thing, then they can conflict with each other.

File: internal/website/content/howto/docstore/actions.md (hunk 6 above)
Comment (jba) [reply internal/website/content/howto/docstore/actions.md]: Done.

File: internal/website/content/howto/docstore/actions.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+Once you have [opened a collection]({{< ref "open-collection.md" >}}), you
+can call action methods on it. We will use `coll` as the variable holding the collection.
+
+Docstore supports six actions on documents:
+
+- `Get` retrieves a document.
+- `Create` creates a new document.
+- `Replace` replaces an existing document.
+- `Put` puts a document whether or not it is already present.
```
Comment (vangent) [inline internal/website/content/howto/docstore/actions.md]: Nit: "already exists" would be slightly clearer (to contrast to `Replace`).

File: internal/website/content/howto/docstore/actions.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+{{< goexample "gocloud.dev/internal/docstore.ExampleCollection_Actions_bulkWrite" >}}
+
+`ActionList` has a fluent API, so you can build and execute a sequence of
+actions in one line of code. Here we `Put` a document and immediately `Get` its new
+contents.
+
+{{< goexample "gocloud.dev/internal/docstore.ExampleCollection_Actions_getAfterWrite" >}}
+
+(If the underlying provider is eventually consistent, the result of the `Get`
```
Comment (vangent) [inline internal/website/content/howto/docstore/actions.md]: Nit: not sure the parens are needed.

File: internal/website/content/howto/docstore/actions.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+set of modifications to document fields, and applies them all atomically. You
+can change the value of a field, increment it, or delete it.
+
+{{< goexample "gocloud.dev/internal/docstore.ExampleCollection_Update" >}}
+
+## Revisions {#docrev}
+
+Docstore maintains a revision for every document. Whenever the document is
+changed, the revision is too. By default, Docstore stores the revision
+in a field named `DocstoreRevision`.
```
Comment (vangent) [inline internal/website/content/howto/docstore/actions.md]: maybe add something about how to change the default? Even if it's just something like "; the `Collection` constructors generally take an `Option` to use a different field."

File: internal/website/content/howto/docstore/actions.md (hunk 7 above)
Comment (jba) [reply internal/website/content/howto/docstore/actions.md]: done

File: internal/website/content/howto/docstore/actions.md (hunk 8 above)
Comment (jba) [reply internal/website/content/howto/docstore/actions.md]: done

File: internal/website/content/howto/docstore/actions.md (hunk 9 above)
Comment (jba) [reply internal/website/content/howto/docstore/actions.md]: done

**Pull request merged**